url3aa01d34ec_katsuster_ememu_tgz-pJ8-net_katsuster_ememu_ui_MainJ8/src/net/katsuster/ememu/arm/DualTimer.java

```diff
@@ -0,0 +1,363 @@
+                case REG_Timer1MIS:
+                    //read only, ignored
+                    break;
+                case REG_Timer1BGLoad:
+                    loadValue[0] = data;
+                    super.writeWord(regaddr, data);
+                    break;
+                case REG_Timer2Load:
+                    loadValue[1] = data;
+                    currentValue[1] = 0;
+                    super.writeWord(regaddr, data);
+                    break;
+                case REG_Timer2Value:
+                    //read only, ignored
+                    break;
+                case REG_Timer2Control:
+                    updateControl(1, regaddr, data);
+                    break;
+                case REG_Timer2IntClr:
+                    rawInt[1] &= ~data;
+                    break;
+                case REG_Timer2RIS:
+                case REG_Timer2MIS:
+                    //read only, ignored
+                    break;
+                case REG_Timer2BGLoad:
+                    loadValue[1] = data;
+                    super.writeWord(regaddr, data);
+                    break;
+                case REG_TimerPeriphID0:
+                case REG_TimerPeriphID1:
+                case REG_TimerPeriphID2:
+                case REG_TimerPeriphID3:
+                case REG_TimerPCellID0:
+                case REG_TimerPCellID1:
+                case REG_TimerPCellID2:
+                case REG_TimerPCellID3:
+                    //read only, ignored
+                    break;
+                default:
+                    super.writeWord(regaddr, data);
+                    break;
+                }
+            }
+        }
+
+        @Override
+        public void run() {
+            while (!shouldHalt()) {
+                //FIXME: 100Hz polling
+                int hz = 100;
+
+                try {
+                    Thread.sleep(1000 / hz);
+
+                    //NOTE: タイマーカウンタのクリアと排他する必要がある
+                    synchronized (this) {
+                        for (int id = 0; id < 2; id++) {
+                            if (!timerEn[id]) {
+                                //hold value
+                                continue;
+                            }
+
+                            int dec = clock / (hz * prescale[id]);
+                            int after = currentValue[id] - dec;
+
+                            if (currentValue[id] == 0 ||
+                                    after == 0 || after > currentValue[id]) {
+                                //0 または 0 以下に達した
+                                currentValue[id] = loadValue[id];
+                                rawInt[id] = 0x1;
+                            } else {
+                                currentValue[id] = after;
+                            }
+                        }
+
+                        intDst.setRaisedInterrupt(isAssert());
+                    }
+                } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)